src/stan/mcmc/hmc/nuts_classic/base_nuts_classic.hpp

```diff
@@ -0,0 +1,249 @@
+        names.push_back("treedepth__");
+        names.push_back("n_leapfrog__");
+        names.push_back("divergent__");
+        names.push_back("energy__");
+      }
+
+      void get_sampler_params(std::vector<double>& values) {
+        values.push_back(this->epsilon_);
+        values.push_back(this->depth_);
+        values.push_back(this->n_leapfrog_);
+        values.push_back(this->divergent_);
+        values.push_back(this->energy_);
+      }
+
+      virtual bool compute_criterion(ps_point& start,
+                                     typename Hamiltonian<Model, BaseRNG>
+                                     ::PointType& finish,
+                                     Eigen::VectorXd& rho) = 0;
+
+      // Returns number of valid points in the completed subtree
+      int build_tree(int depth, Eigen::VectorXd& rho,
+                     ps_point* z_init_parent, ps_point& z_propose,
+                     nuts_util& util,
+                     interface_callbacks::writer::base_writer& writer) {
+        // Base case
+        if (depth == 0) {
+            this->integrator_.evolve(this->z_, this->hamiltonian_,
+                                     util.sign * this->epsilon_,
+                                     writer);
+            rho += this->z_.p;
+
+            if (z_init_parent) *z_init_parent = this->z_;
+            z_propose = this->z_;
+
+            double h = this->hamiltonian_.H(this->z_);
+            if (boost::math::isnan(h))
+              h = std::numeric_limits<double>::infinity();
+
+            util.criterion = util.log_u + (h - util.H0) < this->max_delta_;
+            if (!util.criterion) ++(this->divergent_);
+
+            util.sum_prob += std::min(1.0, std::exp(util.H0 - h));
+            util.n_tree += 1;
+
+            return (util.log_u + (h - util.H0) < 0);
+
+          } else {
+          // General recursion
+          Eigen::VectorXd left_subtree_rho(rho.size());
+          left_subtree_rho.setZero();
+          ps_point z_init(this->z_);
+
+          int n1 = build_tree(depth - 1, left_subtree_rho, &z_init,
+                              z_propose, util, writer);
+
+          if (z_init_parent) *z_init_parent = z_init;
+
+          if (!util.criterion) return 0;
+
+          Eigen::VectorXd right_subtree_rho(rho.size());
+          right_subtree_rho.setZero();
+          ps_point z_propose_right(z_init);
+
+          int n2 = build_tree(depth - 1, right_subtree_rho, 0,
+                              z_propose_right, util, writer);
+
+          double accept_prob = static_cast<double>(n2) /
+            static_cast<double>(n1 + n2);
+
+          if ( util.criterion && (this->rand_uniform_() < accept_prob) )
+            z_propose = z_propose_right;
+
+          Eigen::VectorXd& subtree_rho = left_subtree_rho;
+          subtree_rho += right_subtree_rho;
+
+          rho += subtree_rho;
+
+          util.criterion &= compute_criterion(z_init, this->z_, subtree_rho);
```

Comment:
Be careful with this, because `&=` is bitwise, not logical, so it'd be clearner with an explicit `util.criterion && ...` on the RHS.